[Frontend] Node server artifact handler supports preview and handles both archived and unarchived (gzip/tarball) artifacts (#2172)

eterna2 opened the pull request:
Refer #2172

### Changes
Server:
- swap node modules `tar` with `tar-stream` and `maybe-gunzip` to support tarballs with and w/o compression
- updated artifact endpoints with optional `peek` flag, which will return the first <peek> chars/bytes of the artifact if flag is provided.

UI
- Updated MinioArtifactLink to call artifact endpoint and replace link text with preview.

Automated build for this PR can be found (tag=artifact-preview)
link: https://hub.docker.com/repository/docker/e2forks/ml-pipeline-ui

### Updated 20 mar 2020

Server
- All artifact handlers now support preview with `PreviewStream` (stream first x bytes if peek is provided, other acts like passthrough)
- s3 and minio artifacts now share the common `getObjectStream` func which will inflate/ungzip any compressed artifacts, streams the 1st record if the artifact is a tarball, otherwise stream the raw contents.
- changes to `decodeURIComponents` are reverted  (server.ts only - have not fix frontend)
- added the reference to `tar` magic number offset and chars.

Frontend
- `DetailsTable` now directly renders `MinioArtifactPreview` if the value of the fields is an Argo S3Artifact obj
- Fixed unit tests issue with react hooks by using `mount` instead of `shallow`
- `MinioArtifactLink` is renamed to `MinioArtifactPreview` and now renders a preview of the artifact content
- `MinioArtifactPreview` now accepts an artifact prop (which is an argo S3Artifact obj) because `key` is a protected prop in react. i.e. `<MinioArtifactPreview artifact={s3obj} />` instead of `<MinioArtifactPreview {...s3obj.key} />

TODO: 
- fix front-end issue where `encodeURIComponents` are called multiple times

### Updated 23 Mar 2020

Decoupled frontend and node server codes for easier review. This PR now has the following changes:

- `PreviewStream` to support preview of any data stream
- `maybeTarball` to stream first record of a tarball if the artifact is a tarball, otherwise acts like a passthrough
- minio and s3 artifacts are now handled in the same way with `getObjectStream`
- `getObjectStream` now supports `gunzip-maybe` and `mayberTarball`
- all artifact handlers now supports `PreviewStream` and an additional query string `peek`


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/2992)
<!-- Reviewable:end -->

Bobgy commented:
/ok-to-test

Bobgy commented:
@eterna2 Thank you for the contribution!! Taking a look

Bobgy commented on frontend/src/components/MinioArtifactLink.tsx:
```diff
@@ -1,30 +1,36 @@
-    >
-      {linkText}
+    <a href={url} target={'_blank'} rel={'noreferrer noopener'} title={linkText}>
+      {content}
```
Some opinions about this UX:
1. When UI server doesn't have permission to fetch content (that's a common deployment problem), this should better at least show user the link.
2. The link is already known, while content takes some time to fetch. IMO, we should better show the link before fetching results.

With my two opinions, I'm thinking about the following UX:
```
[link]
```
when loaded, it becomes
```
[link]
content
```

What do you think about this?

eterna2 replied:
Yup this is a good idea. Originally I was thinking of extending it abit more.

i.e. if the payload is sufficiently small, to actually render the json/csv/etc.

But let's keep it simple for now.

Bobgy replied:
SG, we can build more features later

Bobgy commented on frontend/src/components/MinioArtifactLink.tsx:
```diff
@@ -1,30 +1,36 @@
+  const url = generateArtifactUrl(source, bucket, encodedKey);
+  const [content, setContent] = React.useState(linkText);
+
+  React.useEffect(() => {
```
Congratulations! You are the first contributor to use react hooks in this repo. : )

Bobgy replied:
nit: you should return a clean up function to the hook. Without the cleanup, things like
1. first request out
2. props changed
3. second request out
4. second request back -> content set to be consistent with current props
5. first request back -> content set again consistent with outdated props
could happen. 

Although, we won't need to support changing props. We need the clean up function at least to avoid warnings like `React setState called when component is already unmount`

```
// an example simplified fix
useEffect(() => {
  let canceled = false;
  Apis.request().then(content => {
    if (!canceled) setContent(content);
  }
  return () => { canceled = false; };
}
```

eterna2 replied:
Btw I was unable to fix the unit test error.

Any ideas? I checked react and react Dom are of the right version. And everything builds fine.

I suspect it might be `enzyme-adapter-react-16`? But I am not really a react person.

```
MinioArtifactLink › handles s3 artifact
    Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
    1. You might have mismatching versions of React and the renderer (such as React DOM)
    2. You might be breaking the Rules of Hooks
```

Bobgy replied:
Looks like we need to upgrade to latest enzyme: https://github.com/airbnb/enzyme/issues/2011

eterna2 replied:
Don't think it is going to be fixed soon. `useEffect` still not supported.
https://github.com/facebook/react/issues/17321

Bobgy replied:
I see, you can avoid testing it in shallow then. Just use mount.

Bobgy commented on frontend/src/components/MinioArtifactLink.tsx:
```diff
@@ -1,30 +1,36 @@
+  React.useEffect(() => {
+    Apis.readFile({ source, bucket, key }, peek).then(
+      data => setContent(data.length > peek ? `${data} ...` : data),
+      error => setContent(error),
```
Directly showing error message doesn't look good, also there's no way to retry.
Two options I can think of:
* Don't show the preview if it failed
* Show a written error message and log the root cause in console

Bobgy commented on frontend/src/components/MinioArtifactLink.tsx:
```diff
@@ -1,30 +1,36 @@
- * A component that renders an artifact link.
+ * A component that renders a preview to an artifact with a link to the full content.
  */
 const MinioArtifactLink: React.FC<S3Artifact> = s3artifact => {
```
Please rename after the change too

Bobgy commented on frontend/server/handlers/artifacts.ts:
```diff
@@ -140,17 +155,38 @@ function getS3ArtifactHandler(options: { bucket: string; key: string; client: Mi
-function getMinioArtifactHandler(options: { bucket: string; key: string; client: MinioClient }) {
+function getMinioArtifactHandler(
+  options: { bucket: string; key: string; client: MinioClient },
+  peek: number = -1,
```
nit: default to 0?

Bobgy commented on frontend/src/components/MinioArtifactLink.tsx:
```diff
@@ -1,30 +1,36 @@
+
+  React.useEffect(() => {
+    Apis.readFile({ source, bucket, key }, peek).then(
+      data => setContent(data.length > peek ? `${data} ...` : data),
```
It seems your peek behavior is returning at least peek characters. However, if exactly peek characters are returned. How do you know if there's more data?
I'd think you should peek for peek+1 characters, and only show peek characters with the `...`.

eterna2 replied:
Good idea

eterna2 commented on frontend/server/minio-helper.ts:
```diff
@@ -12,7 +12,8 @@
 // limitations under the License.
 import { Stream } from 'stream';
-import * as tar from 'tar';
+import * as tar from 'tar-stream';
```
`tar` couples the deflating/gunzip with extracting the tarball.

I need to search for the issue. Essentially, there is a slightly inconsistency in storing the artifact with Argo in diff storage backend. Some had gzip compression, others dun.

i.e. I recalled I had a bit of issue getting my metadata UI to render properly, because s3 artifact is stored differently from a minio artifact.

So there was a discussion on turning gzip off for everything to ensure consistency. So I need to switch to `tar-stream` and `maybe-gunzip` to handle compressed and uncompressed tarballs.

Let me search for the issue again. It was many months ago.

eterna2 replied:
Ok. I think I recalled more of it. This PR still haven't fix the issue. What we need is a `maybe tarball.`

U notice s3 artifacts are not in archived. While all minio artifacts are tarballs. 

In the past, before I added update the minio client to be able to retrieve from s3 w/o API key. For metadata-ui, we store the data in s3 in our kubeflow op. But retrieve in UI as a minio artifact (with a minio gateway to s3).

However, the API for minio artifact expects a tarball, but in reality our artifact is not.

eterna2 commented:
Made all the changes. Updated `enzyme` but doesn't seems to help. 

Added a screenshot to show how it looks now.
![artifactpreview](https://user-images.githubusercontent.com/1248825/74166530-aca11d00-4c61-11ea-84f7-c7aebe602a45.png)

eterna2 commented:
I was testing in my local machine with the mock server. It works perfect as shown in the screenshot.

However, when i deployed it to my dev cluster, the frontend ui crashed at the `useState` portion. I am still investigating what is wrong. 

But so far I have little clue. 

I have used hooks in my own app, but they works perfectly. But I am not using ezyme thou.

Bobgy commented:
Strange, I don't have much bandwidth to help you troubleshoot right now. I am also using hooks for a recent PR. I will let you know if there are any problems on my side.

Bobgy commented:
I tried using react hooks in https://github.com/kubeflow/pipelines/pull/3043.
I skipped unit testing its behavior in shallow rendering tests, and at least it's not breaking anything. I'd expect using enzyme mount for testing would work. (or you can just use react-testing-library, which I'd prefer we move towards later.)

Also I tried building an image from my PR, there doesn't seem any problems. So that rules out problems simply at using hooks. Could there be other issues in this PR?

Bobgy commented on frontend/src/pages/RunDetails.tsx:
```diff
@@ -326,15 +326,15 @@ class RunDetails extends Page<RunDetailsProps, RunDetailsState> {
                                       title='Input artifacts'
                                       fields={inputArtifacts}
-                                      valueComponent={MinioArtifactLink}
+                                      valueComponent={MinioArtifactPreview}
```
Also here is also problematic too.
https://github.com/Bobgy/pipelines/blob/fe8928926b42a51976e2f421642924a3771afde6/frontend/src/components/DetailsTable.tsx#L106

It should return <MinioArtifactLink {...value} />, not calling the function directly.

This explains why after building an image, your UI crashes. These are unexpected behaviors for react.

eterna2 replied:
👍 i went to properly read up on react. Now I have a better idea, how jsx works.

eterna2 replied:
👍 i went to properly read up on react. Now I have a better idea, how jsx works.

eterna2 commented:
Hey @Bobgy 

I fixed all the issues.

Frontend:
- `DetailsTable` now directly renders `MinioArtifactPreview` if the value of the fields is an Argo S3Artifact obj
- Fixed unit tests issue with react hooks by using `mount` instead of `shallow`
- `MinioArtifactLink` is renamed to `MinioArtifactPreview` and now renders a preview of the artifact content
- `MinioArtifactPreview` now accepts an artifact prop (which is an argo S3Artifact obj) because `key` is a protected prop in react. i.e. `<MinioArtifactPreview artifact={s3obj} />` instead of `<MinioArtifactPreview {...s3obj.key} />

Nodejs:
- S3 and Minio artifact handlers are now made consistent with each other - i.e. both uses the `getObjectStream` function

- `getObjectStream` function will uncompress gzipped/deflated objects, and if the obj is a tarball, returns the 1st object inside the tarball, otherwise it will just return the content of the object.

eterna2 commented on frontend/src/components/MinioArtifactPreview.test.tsx:
```diff
@@ -14,19 +14,19 @@
+describe('MinioArtifactPreview', () => {
   it('handles undefined artifact', () => {
-    expect(MinioArtifactLink(undefined as any)).toMatchSnapshot();
+    expect(MinioArtifactPreview(undefined as any)).toMatchSnapshot();
```
fixed.

eterna2 commented on frontend/src/lib/Utils.tsx:
```diff
@@ -277,8 +277,23 @@ export function generateGcsConsoleUri(gcsUri: string): string | undefined {
+  key: string,
+  peek?: number,
+): string {
+  return encodeURI(
```
yeah. I did investigate before. `/` inside query string is fine as long as it happens after `?`.

But just to be extra sure, I added `encodeURIComponent` back.

eterna2 commented:
I also figured out where the runtime error is. It is not the hooks. But because my limited understanding of jsx (I do mostly webcomponents instead of react.components) and the type check is not guaranteed.

I did not realized that u cannot pass an obj directly inside the react components.

```jsx
<some-component>
{someFoo}
</some-component>
```

`someFoo` in this case must be a react component or a string. It cannot be a javascript object (i.e. it is not automatically converted to a string).

Because I check whether a field value is an argo artifact obj (renders a preview if so, otherwise render the obj as is).

So if the argo artifact obj that is returned is incomplete or wrong (cuz this are io inputs), i will try to render the obj - which will cuz the error.

Bobgy commented:
@eterna2 Awesome! I will take a look when I have some time.

eterna2 commented:
Modified the encode and decide uri components as I realised that multiple encode/decode can yield different outputs.

This creates an issue in the node server as we only decode once and use the key to get the object.

Whereas in the front end, if the key are accidentally encoded twice or more (e.g. it is encoded before u pass to the API generator, which will encode it again), you need to decode the same number of times in order to get back the original string.

So I created 2 function called consistentEncodeURIComponent and consistentDecodeURIComponent which will ensure that the string will only be encoded once, and encoded string will always be decoded completely.

Bobgy commented:
Hi @eterna2, sorry please expect some delay before I can review this. It's been busy recently.

Ark-kun commented:
/retest

Bobgy commented:
Sorry for the delay, I'm looking at this now

Bobgy commented:
A general question first:
The UI always fetches artifact content, peek count really helps UI to be scalable. (not problem until here).
However, the UI server is only supporting real peek for minio client, but not gcs/http client. If user is storing a lot of large artifacts with gcs/http artifacts, it seems UI server will keep downloading too much content. Related code: https://github.com/kubeflow/pipelines/pull/2992/files#diff-fcbaf5299290dcb513dc9d2f3fc67731R224

Bobgy commented on frontend/server/handlers/artifacts.ts:
```diff
@@ -204,7 +221,7 @@ function getGCSArtifactHandler(options: { key: string; bucket: string }) {
                 .trim() + '\n';
             if (i === matchingFiles.length - 1) {
-              res.send(contents);
+              res.send(peek > 0 ? contents.slice(0, peek) : contents);
```
As I mentioned in https://github.com/kubeflow/pipelines/pull/2992#issuecomment-598085681, I'd prefer a proper peek behavior here as well, to avoid scalability issues.

eterna2 replied:
Ok. Might need to research more on gcs sdk, cuz I am not terribly familiar with it. But shld not be an issue.

Bobgy commented on frontend/server/minio-helper.ts:
```diff
@@ -54,24 +56,80 @@ export async function createMinioClient(config: MinioClientOptionsWithOptionalSe
-      reject(err);
-    }
+/**
+ * Checks the magic number of a buffer to see if the mime type is a uncompressed
```
Can you put a reference link here?

eterna2 replied:
done
Add `https://www.gnu.org/software/tar/manual/html_node/Standard.html` to comment.

Bobgy commented on frontend/server/minio-helper.ts:
```diff
@@ -54,24 +56,80 @@ export async function createMinioClient(config: MinioClientOptionsWithOptionalSe
+    { newline: false, maxBuffer: 264 },
+    (data: Buffer, swap: (error?: Error, parser?: Transform) => void) => {
+      if (isTarball(data)) swap(null, extractFirstTarRecordAsStream());
+      else swap(null, new PassThrough());
```
So we also support uncompressed artifacts after this change?
Great work!

eterna2 replied:
Yeah. The flow is
- if compressed, then deflate or gunzip, else just stream as normal
- if tarball, stream 1st record, else just stream everything

Bobgy commented on frontend/server/utils.ts:
```diff
@@ -61,3 +61,32 @@ export function loadJSON<T>(filepath?: string, defaultValue?: T): T | undefined
+ *
+ * @param uri uri string to decode
+ */
+export function consistentDecodeURIComponent(uri: string) {
```
I don't think this is the right way to do it.
What if the URI want to have '%' as payload?
and it can also throw exceptions when decoding like the following
```
$ decodeURIComponent('100%25=0')
"100%=0"
$ decodeURIComponent('100%=0')
VM392:1 Uncaught URIError: URI malformed
    at decodeURIComponent (<anonymous>)
    at <anonymous>:1:1
```

I think server must explicitly know how many times it was actually encoded.... actually, client should encode exactly once.

eterna2 replied:
Yeah, I agree with u. 

But might need some research. From frontend, shld not be an issue as long as we enforce where uri will be encoded.

But on backend side, my issue is I am not sure about the uri I extract from workflow status with k8s API.

How shld we handle that?

Bobgy replied:
That's a good point. I'm not sure about that.

Can we try using the object key directly? (Sounds to me that key shouldn't be encoded, because it was stored as a single field.)
Do you have a case it will be encoded for at least once?

Ark-kun replied:
Let's scope the issue down to valid S3 URIs that can be generated and used by Argo.

eterna2 replied:
let me revert the changes. probably can look at this separately. I will ensure the front-end only encode it once.

Bobgy commented:
/retest

eterna2 commented:
Ok. I fixed most of the stuff for the server-side.

Let me summarize the changes so far:
#### server
- All artifact handlers now support preview with `PreviewStream` (stream first x bytes if peek is provided, other acts like passthrough)

- s3 and minio artifacts now share the common `getObjectStream` func which will inflate/ungzip any compressed artifacts, streams the 1st record if the artifact is a tarball, otherwise stream the raw contents.

- changes to `decodeURIComponents` are reverted  (server.ts only - have not fix frontend)


### TODO: 
- fix front-end issue where `encodeURIComponents` are called multiple times

Ark-kun commented:
/retest

Bobgy commented on frontend/server/utils.ts:
```diff
@@ -63,3 +64,34 @@ export function loadJSON<T>(filepath?: string, defaultValue?: T): T | undefined
+/**
+ * Transform stream that only stream the first X number of bytes.
+ */
+export class PreviewStream extends Transform {
```
Great work! This abstraction makes code easier to reason about!

Bobgy commented on frontend/src/components/DetailsTable.test.tsx:
```diff
@@ -16,8 +16,63 @@
+describe('DetailsFieldValue', () => {
+  it('does render string as SPAN', () => {
+    const tree = shallow(<DetailsFieldValue index={1} fieldname={'key'} fieldvalue={'text'} />);
+    expect(tree).toMatchSnapshot();
```
FYI, new snapshot testing guideline is to always use `toMatchSnapshotIneline()` and avoid taking long snapshots.

If several snapshots share a lot of things in common, take advantage of snapshot-diff to reduce duplication: e.g. https://github.com/kubeflow/pipelines/blob/5162e455a8571c55d6af5a995c835b33e2d8751c/frontend/src/components/viewers/Tensorboard.test.tsx#L61

For details: https://github.com/kubeflow/pipelines/pull/3166

Bobgy commented on frontend/src/components/DetailsTable.tsx:
```diff
@@ -49,62 +50,82 @@ export const css = stylesheet({
+  // if js object, renders a editor
+  if (fieldvalue && typeof fieldvalue === 'object') {
+    return (
+      <div key={index} className={css.row}>
```
```suggestion
      <div className={css.row}>
```
You don't need a key here, it's not a list.

Bobgy replied:
Hmm, actually, why do we need the prop index?

eterna2 replied:
Actually, I have no idea too. It was in the original code, I kept it inside in case of breaking anything thou. 

I reordered the rendering sequence to support the preview.

eterna2 commented on frontend/package-lock.json:
```diff
@@ -6402,7 +6402,8 @@
       "version": "3.11.4",
       "resolved": "https://registry.npmjs.org/@types/js-yaml/-/js-yaml-3.11.4.tgz",
-      "integrity": "sha512-a42nbZzOlxbH3ORDg/brU+zlRQS8mSvZCEY4Xery66/NIA7yX4T9qk6yL9qpw4AhgwzEgc1XIhHRnivQW+zlkw=="
+      "integrity": "sha512-a42nbZzOlxbH3ORDg/brU+zlRQS8mSvZCEY4Xery66/NIA7yX4T9qk6yL9qpw4AhgwzEgc1XIhHRnivQW+zlkw==",
```
actually i removed `@types/js-yaml` from package.json cuz i don't see it used anywhere. But I reverted it anyway.

And we probably want to put in in dev, since everything is compiled to js.

eterna2 commented:
@Bobgy 

Decoupled frontend and node server codes for easier review. (I will make another PR for the frontend).

This PR now only has the following changes:

- `PreviewStream` to support preview of any data stream
- `maybeTarball` to stream first record of a tarball if the artifact is a tarball, otherwise acts like a passthrough
- minio and s3 artifacts are now handled in the same way with `getObjectStream`
- `getObjectStream` now supports `gunzip-maybe` and `mayberTarball`
- all artifact handlers now supports `PreviewStream` and an additional query string `peek`

Bobgy commented:
Hmmm, it happened again, maybe we should avoid using Github suggestion feature. I think google cla bot didn't recognize that commit, because its email says github bot, instead your email.

eterna2 commented:
yeah. fixed it.

Bobgy commented:
/retest

Bobgy commented:
/lgtm
/unhold
